Feature/xx 3 detailed vc

jellocrew opened the pull request.

jellocrew committed:
xx-2. add city VC

jellocrew committed:
xx-2.1 city name to mainVC

jellocrew committed:
xx-2.1_some_fix

jellocrew committed:
xx-2.2 fix empty city, add share

jellocrew committed:
xx-2.3 add constraint (field and tableView)

jellocrew committed:
xx 2.4 add searchBar for city Search

jellocrew committed:
xx-2.5 CitiesStore class, updateBackground, delete in storyboard

jellocrew committed:
xx-2.6 Notification - GetPartOfDay and someFix

jellocrew committed:
xx-3 add some constraint programmatically

jellocrew committed:
xx-3.1 fix: add day textcolor

jellocrew committed:
xx-2.7 - fix getPartOfDay

jellocrew committed:
Merge branch 'feature/xx-2_add_city_vc' into develop

jellocrew committed:
Merge branch 'xx-3_DetailedVC' into develop

# Conflicts:
#	Weather_app_mvvm_test/Model/GetPartOfDay.swift

dkorneev commented on Weather_app_mvvm_test/Model/GetPartOfDay.swift:
```diff
@@ -0,0 +1,49 @@
+    var backgroundColor: UIColor?
+    var textColor: UIColor?
+    
+    func updateBackgroundColor(){
```
Я бы написал геттер для каждого проперти, в котором время дня определял бы "на лету" и возвращал нужный цвет. Это не так эффективно с точки зрения потребления процессорного времени (один и тот же код выполняется каждый раз при обращении к свойству), но гораздо удобнее для пользователя этого класса. Так ты не забудешь в нужном месте вызвать `updateBackgroundColor()` и у тебя исключена ошибка в этом плане.
Ну а подобной неэффективностью на практике можно пренебречь.

dkorneev commented on Weather_app_mvvm_test/Model/GetPartOfDay_BACKUP_2665.swift:
```diff
@@ -0,0 +1,65 @@
+enum PartOfDay {
```
Что за бекап?) В гите вся история файлов есть, если что и всегда можно вытащить старый файл или предыдущую его версию.

jellocrew replied:
Это я ветки в порядок приводил. tower мне создал много версий в stashes) сейчас почистил все.

dkorneev replied:
что за tower?

dkorneev commented:
Какая-то путаница с файлами GetPartOfDay, в проекте должен лежать только один актуальный файл. Все остальное удаляем. Старые версии и удаленные файлы можно смотреть в истории гита. Для этого он и используется.

dkorneev commented on Weather_app_mvvm_test/Model/LocationManager.swift:
```diff
@@ -0,0 +1,53 @@
+    let coordinates: CLLocationCoordinate2D
+}
+
+class LocationManager: NSObject {
```
Почему NSObject?

jellocrew replied:
этот класс не нужен уже, удалю, он для textfied с поиском города нужен был.

dkorneev commented:
Большие картинки (не иконки) рекомендую в jpg перегонять, чтобы не увеличивать размер бинарника. Сейчас у тебя лежат несколько картинок по 200КБ, довольно большие.

dkorneev commented on Weather_app_mvvm_test/View/AddCityVC.swift:
```diff
@@ -0,0 +1,85 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        
+        let searchResult = searchResults[indexPath.row]
+        let cell = UITableViewCell(style: .subtitle, reuseIdentifier: nil)
```
UITableViewCell создается каждый раз новая, а не реюзается через tableView.deque...

Фишка tableView в переиспользовании ячеек. При скролле, если она не создает каждый раз ячейку, а переконфигурирует уже созданную, то интерфейс работает плавно. На современных девайсах, для простых ячеек и такие неэффективные способы могут работать нормально, до определенной сложности ячейки. Но все равно использовать нужно подход с реюзом.

Почитай доку по UITableView, это один из основных компонентов в UIKit и про реюз ячеек там или в UICollectionView вероятно спросят на собеседовании.

dkorneev commented on Weather_app_mvvm_test/View/AddCityVC.swift:
```diff
@@ -0,0 +1,85 @@
+        var listCitiesArray = store.loadCities()
+        listCitiesArray.append(cityName)
+        store.saveCities(listCitiesArray: listCitiesArray)
+        NotificationCenter.default.post(name: .userDefaultsIsUpdate, object: nil, userInfo: nil)
```
Я бы отправлял уведомление из стора, а не из контроллера. Чтобы при каждой записи в хранилище, наблюдатели уведомлялись, вне зависимости откуда запись была произведена.

dkorneev commented on Weather_app_mvvm_test/View/AddCityVC.swift:
```diff
@@ -0,0 +1,85 @@
+    
+    //MARK: - Flow func
+    func searchBar(_ searchBar: UISearchBar, textDidChange searchText: String) {
+        searchCompleter.queryFragment = searchText
```
Думал ты юзаешь какой-то сторонний сервис для получения городов. А с этим классом они сложную часть за тебя сделали, удобно))

dkorneev replied:
<img width="776" alt="image" src="https://user-images.githubusercontent.com/3111175/176772613-ea33c87a-913d-4acc-9f93-608731a72006.png">

dkorneev commented:
Вообще прикольно получается.

jellocrew committed:
xx-3 add collectionView

jellocrew committed:
xx-3.2 add daily and hourly forecast

replaced all weather models, add collection view and table view with xib for daily and hourly forecast

jellocrew committed:
xx-3.2.1 some fix

jellocrew committed:
xx-3.2.2 some fix

jellocrew committed:
xx-2.8 dequeueReusableCell + xib

jellocrew committed:
Merge branch 'feature/xx-2_add_city_vc' into develop

jellocrew committed:
Merge branch 'develop' into feature/xx-3_DetailedVC

# Conflicts:
#	Weather_app_mvvm_test.xcodeproj/project.pbxproj
#	Weather_app_mvvm_test/View/AddCityVC.swift
#	Weather_app_mvvm_test/View/MainCells/CitySearchTableViewCell.swift
#	Weather_app_mvvm_test/View/MainCells/CitySearchTableViewCell.xib

jellocrew committed:
xx-3.3 merge conflict fix

jellocrew committed:
xx-3.3.1 some fix
